### PR TITLE
Support 3p APIs in chrome.identity without Google API keys

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/15754

This patch replaces the chrome.identity API, which relies on 1p APIs,
with a version that relies on 3p APIs, necessary for some extensions
to work Brave after chrome.identity changed to only support 1p APIs.

Applying upstream patch as-is https://crrev.com/c/2704571

Security / privacy review https://github.com/brave/security/issues/947

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Download/ install an extension with 3p authorization; @spylogsster used https://chrome.google.com/webstore/detail/drive-sample/jpabeekbjicamajjcfejnochhmlbpgjh/related?pli=1
- Login to extension using Google login and verify you can login

### CODE REVIEW

File: chromium_src/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "base/time/time.h"
+
+#define FromSeconds \
+  Max().is_max() ? base::Seconds(time_to_live_seconds) : base::Seconds
+#define FALLTHROUGH [[fallthrough]]
```
Comment (bsclifton) [inline chromium_src/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc:10]: Curious about this one - why did it need to be defined? Is `FALLTHROUGH` not defined anywhere in Chromium source?

Comment (spylogsster) [reply chromium_src/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc:10]: yes, it throws errors

**Review by bsclifton — APPROVED** [review]
LGTM 👍

**Comment by bsclifton** [comment]
Created a security / privacy review just to make sure this is OK - let's please not merge this until that is completed
https://github.com/brave/security/issues/947

File: chromium_src/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc (hunk 1 above)
Comment (bsclifton) [reply chromium_src/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc:10]: OK cool - can you add a comment for that? 😄

**Review by kdenhartog — COMMENTED** [review]
Just to confirm, this web OAuth flow is only used by G Suite extensions once a user has added them and it won't trigger if the user hasn't already added these extensions?

**Comment by spylogsster** [comment]
> Just to confirm, this web OAuth flow is only used by G Suite extensions once a user has added them and it won't trigger if the user hasn't already added these extensions?

this can be used by any extension that uses chrome.identity api. The changes affect only chrome.identity api implementation and cannot be used outside of it

**Review by kdenhartog — APPROVED** [review]
> this can be used by any extension that uses chrome.identity api. The changes affect only chrome.identity api implementation and cannot be used outside of it

Ok, in that case I don't see any concerns here. 

Sorry, I realize now I wasn't as clear about the goal I was checking for which was to see if we would unintentionally allow network requests to non-brave servers with this functionality if the user hasn't modified anything after a fresh install. Based on my understanding of this, I don't see how that would be possible since these additional APIs are only exposed to extensions which requires the user to install an extension so it wouldn't be a concern here.

I'm good with this from a security perspective, and can we [update the docs](https://github.com/brave/brave-browser/wiki/Allow-Google-login---Third-Parties-and-Extensions#extensions) to reflect this has been fixed now as well?

**Pull request merged**